File: elixir/23-MergeKSortedLists/main.exs
defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  def merge_k_lists([]) do
    nil
  end

  def merge_k_lists([list]) do
    list
  end

  @spec merge_k_lists(lists :: [ListNode.t | nil]) :: ListNode.t | nil
  def merge_k_lists(lists) do
    [first | rest] = lists
    Enum.reduce(rest, first, fn e, acc -> merge_two_lists(e, acc) end)
  end

  @spec merge_two_lists(list1 :: ListNode.t | nil, list2 :: ListNode.t | nil) :: ListNode.t | nil
  defp merge_two_lists(list1, list2) when list1.val < list2.val  do
    %ListNode{val: list1.val, next: merge_two_lists(list1.next, list2)}
  end

  defp merge_two_lists(list1, list2) when list1.val >= list2.val do
    %ListNode{val: list2.val, next: merge_two_lists(list1, list2.next)}
  end

  defp merge_two_lists(list1, nil) do
    list1
  end

  defp merge_two_lists(nil, list2) do
    list2
  end
end

defmodule Tests do
  defp list_to_string(list) when list == nil do
    ""
  end

  defp list_to_string(list) when list == [] do
    ""
  end

  defp list_to_string(list) do
    "#{list.val} " <> list_to_string(list.next)
  end

  def example1 do
    list1 = %ListNode{val: 1, next: %ListNode{val: 10, next: %ListNode{val: 400}}}
    list2 = %ListNode{val: 2, next: %ListNode{val: 30, next: %ListNode{val: 500}}}
    list3 = %ListNode{val: 40, next: %ListNode{val: 70, next: %ListNode{val: 600}}}
    IO.puts(list_to_string(Solution.merge_k_lists([list1, list2, list3])))
  end

  def example2 do
    IO.puts(list_to_string(Solution.merge_k_lists([])))
  end
end

Tests.example1
Tests.example2

File: cpp/23-MergeKSortedLists/main.cpp
#include <vector>
#include "../LeetCode.h"

using namespace std;

class Solution {
private:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) 
    {
        if(!list1) return list2;
        if(!list2) return list1;

        auto head = grab(list1, list2);
        auto next = head;

        while(list1 && list2)
        {
            next->next = grab(list1, list2);
            next = next->next;
        }

        next->next = (list1 ? list1 : list2);

        return head;
    }

    ListNode *grab(ListNode *&list1, ListNode *&list2)
    {
        ListNode *next = nullptr;
        if(list1->val < list2->val)
        {
            next = list1;
            list1 = list1->next;
        }
        else
        {
            next = list2;
            list2 = list2->next;
        }

        return next;
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) 
    {
        if(lists.empty()) return nullptr;    

        auto next = lists.begin();
        ListNode *merged = *next++;

        for(; next != lists.end(); ++next)
        {
            merged = mergeTwoLists(merged, *next);
        }

        return merged;
    }
};

void test1()
{
    auto l1 = makeList({1, 4, 5});
    auto l2 = makeList({1, 3, 4});
    auto l3 = makeList({2, 6});

    Solution s;
    vector<ListNode*> lists = {l1, l2, l3};
    auto merged = s.mergeKLists(lists);
    print(merged);
}

void test2()
{
    Solution s;
    vector<ListNode*> lists;
    auto merged = s.mergeKLists(lists);
    print(merged);
}

int main()
{
    test1();
    test2();
    return 0;
}

